File: IFATCCompanion/AirportSurface/SurfaceConfidenceEvaluator.swift
import Foundation

/// Grades an airport **dataset's** confidence by combining normalized-feature quality
/// with the connectivity of the derived surface graph. (Route confidence is graded
/// separately by `TaxiRouteEngine`, which also factors the aircraft snap and crossings.)
///
/// High confidence requires connected taxiway geometry, taxiway names/references, clear
/// runway geometry, and reliable holding positions. Medium tolerates some inferred
/// holds / missing names. Low means disconnected or largely unnamed geometry.
/// Unavailable means there is not enough to route on at all.
enum SurfaceConfidenceEvaluator {

    static func datasetConfidence(model: AirportSurfaceModel, graph: SurfaceGraph) -> SurfaceConfidence {
        guard model.hasUsableGeometry, !graph.edges.isEmpty else { return .unavailable }

        let namedFraction = graph.namedEdgeFraction
        let connected = graph.componentCount <= 1
        let hasMappedHolds = model.holdingPositions.contains { !$0.inferred }
        let hasRunways = !model.runways.isEmpty && !model.runwayEnds.isEmpty
        let hasParking = !model.parkingPositions.isEmpty

        if namedFraction >= 0.6 && connected && hasMappedHolds && hasRunways {
            return .high
        }
        if (namedFraction >= 0.3 || hasMappedHolds) && hasRunways && (connected || hasParking) {
            return .medium
        }
        return .low
    }
}

File: Android/core/src/main/kotlin/com/h3consultingpartners/ifatccompanion/core/surface/routing/SurfaceConfidenceEvaluator.kt
package com.h3consultingpartners.ifatccompanion.core.surface.routing

import com.h3consultingpartners.ifatccompanion.core.surface.AirportSurfaceModel
import com.h3consultingpartners.ifatccompanion.core.surface.SurfaceConfidence

/**
 * Grades an airport **dataset's** confidence by combining normalized-feature quality
 * with the connectivity of the derived surface graph. (Route confidence is graded
 * separately by [TaxiRouteEngine], which also factors the aircraft snap and crossings.)
 *
 * High confidence requires connected taxiway geometry, taxiway names/references, clear
 * runway geometry, and reliable holding positions. Medium tolerates some inferred
 * holds / missing names. Low means disconnected or largely unnamed geometry.
 * Unavailable means there is not enough to route on at all.
 *
 * Ported from `IFATCCompanion/AirportSurface/SurfaceConfidenceEvaluator.swift`.
 */
object SurfaceConfidenceEvaluator {

    fun datasetConfidence(model: AirportSurfaceModel, graph: SurfaceGraph): SurfaceConfidence {
        if (!model.hasUsableGeometry || graph.edges.isEmpty()) return SurfaceConfidence.UNAVAILABLE

        val namedFraction = graph.namedEdgeFraction
        val connected = graph.componentCount <= 1
        val hasMappedHolds = model.holdingPositions.any { !it.inferred }
        val hasRunways = model.runways.isNotEmpty() && model.runwayEnds.isNotEmpty()
        val hasParking = model.parkingPositions.isNotEmpty()

        if (namedFraction >= 0.6 && connected && hasMappedHolds && hasRunways) {
            return SurfaceConfidence.HIGH
        }
        if ((namedFraction >= 0.3 || hasMappedHolds) && hasRunways && (connected || hasParking)) {
            return SurfaceConfidence.MEDIUM
        }
        return SurfaceConfidence.LOW
    }
}
